medea/src/app/Widgets/Dataflow/dataflowdialog.cpp

```diff
@@ -441,35 +423,64 @@ DefaultEntityContainer& DataflowDialog::constructContainerInstanceItem(Container
  * @throws std::invalid_argument
  * @return
  */
-ComponentInstance& DataflowDialog::constructComponentInstanceItem(ComponentInstanceData* comp_inst, Pulse::View::DefaultEntityContainer* parent)
+ComponentInstance& DataflowDialog::constructComponentInstanceItem(ComponentInstanceData& comp_inst, Pulse::View::DefaultEntityContainer* parent)
 {
-    checkPrecondition(comp_inst, "component instance");
-    return *(new ComponentInstance(comp_inst->getName(), comp_inst->getType(), parent));
+    auto comp_inst_item = new ComponentInstance(comp_inst.getName(), comp_inst.getType(), parent);
+    QHash<QString, PortInstance*> port_instances;
+    for (const auto& port : comp_inst.getPortInstanceData()) {
+        checkNotNull(port, "port instance data");
+        auto& port_item = constructPortInstanceItem(*port);
+        port_instance_cache_.insert(port->getGraphmlID(), &port_item);
+        comp_inst_item->add(&port_item);
+    }
+    for (const auto& worker : comp_inst.getWorkerInstanceData()) {
+        checkNotNull(worker, "worker instance data");
+        auto& worker_item = constructWorkerInstanceItem(*worker);
+        auto icon_size = Defaults::primary_icon_size * 0.75;
+        worker_item.setPrimaryIconSize(icon_size.width(), icon_size.height());
+        comp_inst_item->add(&worker_item);
+    }
+    return *comp_inst_item;
 }
 
 /**
  * @brief DataflowDialog::constructPortInstanceItem
  * @param port_inst
- * @throws std::invalid_argument
  * @return
  */
-PortInstance& DataflowDialog::constructPortInstanceItem(PortInstanceData* port_inst)
+PortInstance& DataflowDialog::constructPortInstanceItem(PortInstanceData& port_inst)
```

Comment:
This shouldn't be returning a reference for manually allocated memory - it will be very easy to cause a memory leak if the address isn't stored in the future, which isn't particularly common practice for references.

Having a reference parameter coming into the function is a good change, but in this case the return value needs to be a pointer.